File: frontend/src/types/scan.ts
export type ScanStatus = 'pending' | 'running' | 'completed' | 'failed';
export type ScanType = 'hashtag' | 'url';
export type TimeRange = '12h' | '24h' | '48h' | '7d';

export interface EngagementMetrics {
  likes: number;
  comments: number;
  saves: number;
  shares: number;
}

export interface ViralPost {
  id: number;
  instagram_post_id: string;
  instagram_url: string | null;
  post_type: string | null;
  thumbnail_url: string | null;
  creator_username: string | null;
  creator_follower_count: number | null;
  engagement: EngagementMetrics;
  viral_score: number | null;
  post_age_hours: number | null;
}

export interface ScanTriggerResponse {
  scan_id: number;
  status: ScanStatus;
  message: string;
}

export interface ScanResponse {
  scan_id: number;
  status: ScanStatus;
  scan_type: ScanType;
  time_range: TimeRange | null;
  target_url: string | null;
  created_at: string;
  completed_at: string | null;
  error_message: string | null;
  results: ViralPost[];
}

export interface ScanHistoryItem {
  scan_id: number;
  status: ScanStatus;
  scan_type: ScanType;
  time_range: TimeRange | null;
  created_at: string;
  completed_at: string | null;
  post_count: number;
}

File: backend/app/schemas/scan.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class ScanRequest(BaseModel):
    """Request body for triggering a time-range viral content scan."""
    time_range: str = Field(
        ...,
        pattern=r"^(12h|24h|48h|7d)$",
        description="Time range for viral post discovery: 12h, 24h, 48h, or 7d",
    )


class AnalyzeUrlRequest(BaseModel):
    """Request body for analyzing a specific Instagram post URL."""
    instagram_url: str = Field(
        ...,
        description="Full Instagram post URL (e.g. https://www.instagram.com/p/SHORTCODE/)",
    )


class EngagementResponse(BaseModel):
    likes: int
    comments: int
    saves: int
    shares: int


class ViralPostResponse(BaseModel):
    id: int
    instagram_post_id: str
    instagram_url: Optional[str]
    post_type: Optional[str]
    thumbnail_url: Optional[str]     # S3 URL if cached; falls back to Instagram URL
    creator_username: Optional[str]
    creator_follower_count: Optional[int]
    engagement: EngagementResponse
    viral_score: Optional[float]
    post_age_hours: Optional[float]

    class Config:
        from_attributes = True


class ScanResponse(BaseModel):
    scan_id: int
    status: str  # pending, running, completed, failed
    scan_type: str
    time_range: Optional[str]
    target_url: Optional[str]
    created_at: datetime
    completed_at: Optional[datetime]
    error_message: Optional[str]
    results: List[ViralPostResponse] = []

    class Config:
        from_attributes = True


class ScanTriggerResponse(BaseModel):
    scan_id: int
    status: str
    message: str


class ScanHistoryItem(BaseModel):
    scan_id: int
    status: str
    scan_type: str
    time_range: Optional[str]
    created_at: datetime
    completed_at: Optional[datetime]
    post_count: int

    class Config:
        from_attributes = True
